Sources/Harmony/Data/DataSource/GenericNetworkDataSource.swift

```diff
@@ -0,0 +1,100 @@
+//
+// Copyright 2022 Mobile Jazz SL
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+import Alamofire
+
+open class GetNetworkDataSource<T: Decodable>: GetDataSource {
+
+    private var url: String
+    private var session: Session
+
+    private enum EntityCount {
+        case multiple
+        case single
+    }
+    
+    public init(url: String, session: Session) {
+        self.url = url
+        self.session = session
+    }
+    
+    @discardableResult
+    open func getAll(_ query: Query) -> Future<[T]> {
+        return execute(query, type: .multiple)
+    }
+    
+    @discardableResult
+    open func get(_ query: Query) -> Future<T> {
+        
+        let futureList = execute(query, type: .single)
+        
+        guard let firstEntity = try? futureList.result.get().first else {
+            return Future()
+        }
+        
+        return Future(firstEntity)
+    }
+    
+    private func execute(_ query: Query, type: EntityCount) -> Future<[T]> {
```

Comment:
For me this implementation is confusing. You are decoding everything into a list but then just taking the first value when this is not a list.

At this moment I don't have a better recommendation but may be worth to think how to restructure this part to avoid this.

Reply:
That makes sense but nothing comes to mind either to be able to reuse the code in execute function. Otherwise i'll need to repeat that piece of code to fetch single entity and multiple entities... I'll think about that while continuing to implement unit tests

Reply:
You can use a generic method to perform some logic on several types, that way you can also avoid having the type EntityCount to differenciate between a single value and a list.

`func execute<K: Codable>(_ query: Query) -> Future<K>`

Then you can call that method in a context where `K` is inferred to be `T` and in another context where `K` is `[T]`.  The final result would look like in the example below.

`@discardableResult
    open func getAll(_ query: Query) -> Future<[T]> {
        return execute(query)
    }
    
    @discardableResult
    open func get(_ query: Query) -> Future<T> {
        execute(query)
    }`

Reply:
Thank you

Reply:
Done